

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>biollm.base.load_scbert &mdash; BioLLM 0.1.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=a58bc63e"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            BioLLM
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../Cell_embedding.html">Evaluation of the clustering of the embed cell.</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../GRN.html">GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Annotation.html">Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../drug_task.html">Drug Annotation Task Documentation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">BioLLM Modules</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">References:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../download.html">Download the scFM pretrain model and configs</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">BioLLM</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">biollm.base.load_scbert</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for biollm.base.load_scbert</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># coding: utf-8</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">@author: Ping Qiu  qiuping1@genomics.cn</span>
<span class="sd">@last modified by: Ping Qiu</span>
<span class="sd">@file: load_scbert.py</span>
<span class="sd">@time: 2024/3/11 15:02</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">biollm.base.load_llm</span> <span class="kn">import</span> <span class="n">LoadLlm</span>
<span class="kn">from</span> <span class="nn">biollm.repo.scbert.performer_pytorch.performer_pytorch</span> <span class="kn">import</span> <span class="n">PerformerLM</span>
<span class="kn">from</span> <span class="nn">biollm.repo.scgpt.tokenizer.gene_tokenizer</span> <span class="kn">import</span> <span class="n">GeneVocab</span>
<span class="kn">from</span> <span class="nn">biollm.dataset.scbert_dataset</span> <span class="kn">import</span> <span class="n">SCDataset</span><span class="p">,</span> <span class="n">make_scbert_adata</span>
<span class="kn">from</span> <span class="nn">torch.utils.data.sampler</span> <span class="kn">import</span> <span class="n">SequentialSampler</span><span class="p">,</span> <span class="n">RandomSampler</span>
<span class="kn">from</span> <span class="nn">torch.utils.data.distributed</span> <span class="kn">import</span> <span class="n">DistributedSampler</span>
<span class="kn">from</span> <span class="nn">torch.utils.data</span> <span class="kn">import</span> <span class="n">DataLoader</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="kn">import</span> <span class="n">StratifiedShuffleSplit</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">issparse</span>
<span class="kn">import</span> <span class="nn">pickle</span> <span class="k">as</span> <span class="nn">pkl</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>


<div class="viewcode-block" id="LoadScbert">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert">[docs]</a>
<span class="k">class</span> <span class="nc">LoadScbert</span><span class="p">(</span><span class="n">LoadLlm</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    LoadScbert is a class for loading and managing the SCBERT model.</span>
<span class="sd">    This class inherits from LoadLlm and provides functionalities such as loading pretrained models,</span>
<span class="sd">    generating gene and cell embeddings, and preparing data for training.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        num_tokens (int): Number of gene express bins.</span>
<span class="sd">        max_seq_len (int): Maximum sequence length for input data.</span>
<span class="sd">        use_g2v (bool): Indicator to use gene-to-vector (g2v) embeddings.</span>
<span class="sd">        g2v_file (str): Path to the file containing g2v embeddings.</span>
<span class="sd">        vocab (GeneVocab): Vocabulary object for gene-to-index mapping.</span>
<span class="sd">        model (PerformerLM): Loaded SCBERT model.</span>
<span class="sd">        gene2idx (dict): Mapping from gene names to indices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes LoadScbert with specific model settings and loads necessary components.</span>

<span class="sd">        Args:</span>
<span class="sd">            args (Namespace): Arguments object with settings for model, device, and file paths.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">LoadScbert</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__int__</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_tokens</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">bin_num</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_seq_len</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">max_seq_len</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">use_g2v</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">use_g2v</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">g2v_file</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">g2v_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vocab</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_vocab</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_model</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene2idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gene2idx</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">init_model</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>

<div class="viewcode-block" id="LoadScbert.load_model">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.load_model">[docs]</a>
    <span class="k">def</span> <span class="nf">load_model</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads the SCBERT model architecture with Performer-based attention.</span>

<span class="sd">        Returns:</span>
<span class="sd">            PerformerLM: The initialized SCBERT model.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">PerformerLM</span><span class="p">(</span>
            <span class="n">num_tokens</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">num_tokens</span><span class="p">,</span>
            <span class="n">dim</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
            <span class="n">depth</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span>
            <span class="n">max_seq_len</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">max_seq_len</span><span class="p">,</span>
            <span class="n">heads</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
            <span class="n">local_attn_heads</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">g2v_position_emb</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">use_g2v</span><span class="p">,</span>
            <span class="n">g2v_file</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">g2v_file</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">model</span></div>


<div class="viewcode-block" id="LoadScbert.get_gene_embedding">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.get_gene_embedding">[docs]</a>
    <span class="k">def</span> <span class="nf">get_gene_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates embeddings for a set of gene IDs.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_ids (torch.Tensor): Tensor of gene IDs to embed.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: Gene embeddings as a NumPy array.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;start to get gene embedding!&#39;</span><span class="p">)</span>
        <span class="n">emb</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">pos_emb</span><span class="o">.</span><span class="n">emb</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;start to get gene embedding!&#39;</span><span class="p">)</span>
        <span class="n">gene_emb</span> <span class="o">=</span> <span class="n">emb</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;gpu used: &#39;</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">memory_allocated</span><span class="p">())</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="p">:</span>
            <span class="n">total_gpu</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">memory_allocated</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="o">.</span><span class="n">log</span><span class="p">({</span><span class="s1">&#39;GPU&#39;</span><span class="p">:</span> <span class="n">total_gpu</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">gene_emb</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span></div>


<div class="viewcode-block" id="LoadScbert.get_gene2idx">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.get_gene2idx">[docs]</a>
    <span class="k">def</span> <span class="nf">get_gene2idx</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves gene-to-index mapping from vocabulary.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: Dictionary mapping gene names to indices.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">vocab</span><span class="o">.</span><span class="n">get_stoi</span><span class="p">()</span></div>


<div class="viewcode-block" id="LoadScbert.load_vocab">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.load_vocab">[docs]</a>
    <span class="k">def</span> <span class="nf">load_vocab</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads vocabulary from a file, containing mappings of gene names to indices.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GeneVocab: Vocabulary object for gene-to-index mapping.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">vocab</span> <span class="o">=</span> <span class="n">GeneVocab</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">vocab_file</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">vocab</span></div>


<div class="viewcode-block" id="LoadScbert.freezon_model">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.freezon_model">[docs]</a>
    <span class="k">def</span> <span class="nf">freezon_model</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">keep_layers</span><span class="o">=</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Freezes the model layers except for specific normalization and selected layers,</span>
<span class="sd">        which can be fine-tuned.</span>

<span class="sd">        Args:</span>
<span class="sd">            keep_layers (list): List of layers to keep unfrozen for fine-tuning.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">model_param_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">ft_param_count</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">parameters</span><span class="p">():</span>
            <span class="n">model_param_count</span> <span class="o">+=</span> <span class="n">param</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span>
            <span class="n">param</span><span class="o">.</span><span class="n">requires_grad</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">parameters</span><span class="p">():</span>
            <span class="n">param</span><span class="o">.</span><span class="n">requires_grad</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">ft_param_count</span> <span class="o">+=</span> <span class="n">param</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">keep_layers</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">param</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">performer</span><span class="o">.</span><span class="n">net</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">named_parameters</span><span class="p">():</span>
                <span class="n">param</span><span class="o">.</span><span class="n">requires_grad</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">ft_param_count</span> <span class="o">+=</span> <span class="n">param</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total pretrain-model Encoder Params </span><span class="si">{</span><span class="n">model_param_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The pretrain_model Encoder Params for training in finetune after freezon: </span><span class="si">{</span><span class="n">ft_param_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="LoadScbert.preprocess_adata">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.preprocess_adata">[docs]</a>
    <span class="k">def</span> <span class="nf">preprocess_adata</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Preprocesses AnnData object for SCBERT, ensuring compatibility with model input requirements.</span>

<span class="sd">        Args:</span>
<span class="sd">            adata (AnnData): Single-cell data object.</span>

<span class="sd">        Returns:</span>
<span class="sd">            AnnData: Preprocessed AnnData object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">idx2gene</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2idx</span><span class="p">[</span><span class="n">k</span><span class="p">],</span> <span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene2idx</span><span class="p">])</span>
        <span class="n">ref_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">idx2gene</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">idx2gene</span><span class="p">))]</span>
        <span class="k">if</span> <span class="s1">&#39;do_preprocess&#39;</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;the adata was already preprocessed, pass the step!&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">adata</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="n">make_scbert_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">ref_genes</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;do_preprocess&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="n">adata</span></div>


<div class="viewcode-block" id="LoadScbert.load_dataset">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.load_dataset">[docs]</a>
    <span class="k">def</span> <span class="nf">load_dataset</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">split_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">label_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">do_preprocess</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Prepares dataset for training or evaluation, with an option to split into training and validation.</span>

<span class="sd">        Args:</span>
<span class="sd">            adata (AnnData): Single-cell data to process.</span>
<span class="sd">            split_data (bool): Flag to split the dataset into train/validation sets.</span>
<span class="sd">            label_key (str): Column in `adata.obs` containing labels.</span>
<span class="sd">            do_preprocess (bool): Flag to preprocess `adata` if needed.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Tuple[SCDataset, Optional[SCDataset]]: Training and validation datasets.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">do_preprocess</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">preprocess_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">label_key</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">label_dict</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">label_key</span><span class="p">]),</span> <span class="n">return_inverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">label</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span> <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span> <span class="k">else</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
        <span class="k">if</span> <span class="n">split_data</span><span class="p">:</span>
            <span class="n">sss</span> <span class="o">=</span> <span class="n">StratifiedShuffleSplit</span><span class="p">(</span><span class="n">n_splits</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">test_size</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">2021</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">index_train</span><span class="p">,</span> <span class="n">index_val</span> <span class="ow">in</span> <span class="n">sss</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">label</span><span class="p">):</span>
                <span class="n">data_train</span><span class="p">,</span> <span class="n">label_train</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">index_train</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">label</span><span class="p">[</span><span class="n">index_train</span><span class="p">]</span>
                <span class="n">data_val</span><span class="p">,</span> <span class="n">label_val</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">index_val</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">label</span><span class="p">[</span><span class="n">index_val</span><span class="p">]</span>
                <span class="n">train_dataset</span> <span class="o">=</span> <span class="n">SCDataset</span><span class="p">(</span><span class="n">data_train</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">bin_num</span><span class="p">,</span> <span class="n">label_train</span><span class="p">)</span>
                <span class="n">test_dataset</span> <span class="o">=</span> <span class="n">SCDataset</span><span class="p">(</span><span class="n">data_val</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">bin_num</span><span class="p">,</span> <span class="n">label_val</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">train_dataset</span><span class="p">,</span> <span class="n">test_dataset</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">train_dataset</span> <span class="o">=</span> <span class="n">SCDataset</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">bin_num</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">label</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">train_dataset</span><span class="p">,</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="LoadScbert.get_dataloader">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.get_dataloader">[docs]</a>
    <span class="k">def</span> <span class="nf">get_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">drop_last</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">random_sample</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates a DataLoader for batch processing during model training or evaluation.</span>

<span class="sd">        Args:</span>
<span class="sd">            dataset (SCDataset): The dataset to load.</span>
<span class="sd">            drop_last (bool): Flag to drop the last incomplete batch.</span>
<span class="sd">            num_workers (int): Number of worker threads for loading data.</span>
<span class="sd">            random_sample (bool): Flag to enable random sampling.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataLoader: DataLoader for the provided dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">distributed</span><span class="p">:</span>
            <span class="n">sampler</span> <span class="o">=</span> <span class="n">DistributedSampler</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="n">random_sample</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sampler</span> <span class="o">=</span> <span class="n">SequentialSampler</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">random_sample</span> <span class="k">else</span> <span class="n">RandomSampler</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
        <span class="n">data_loader</span> <span class="o">=</span> <span class="n">DataLoader</span><span class="p">(</span>
            <span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">,</span>
            <span class="n">batch_size</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span>
            <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">drop_last</span><span class="o">=</span><span class="n">drop_last</span><span class="p">,</span>
            <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">,</span>
            <span class="n">sampler</span><span class="o">=</span><span class="n">sampler</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">data_loader</span></div>


<div class="viewcode-block" id="LoadScbert.get_gene_expression_embedding">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.get_gene_expression_embedding">[docs]</a>
    <span class="k">def</span> <span class="nf">get_gene_expression_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">pool</span><span class="o">=</span><span class="s1">&#39;mean&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Obtains gene expression embeddings by pooling the model’s cell-level embeddings.</span>

<span class="sd">        Args:</span>
<span class="sd">            adata (AnnData): Single-cell data.</span>
<span class="sd">            pool (str): Pooling method, either &#39;mean&#39; or &#39;sum&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: Gene expression embeddings as a NumPy array.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;start to get gene expression!&quot;</span><span class="p">)</span>
        <span class="n">dataset</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">split_data</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">data_loader</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_dataloader</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">random_sample</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">(),</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">amp</span><span class="o">.</span><span class="n">autocast</span><span class="p">(</span><span class="n">enabled</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
            <span class="n">pool_emb</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2idx</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">embsize</span><span class="p">))</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data_loader</span><span class="p">):</span>
                <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
                <span class="n">cell_encode_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">return_encodings</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># [batch size, max_seq_len, dim]</span>
                <span class="n">cell_encode_x</span> <span class="o">=</span> <span class="n">cell_encode_x</span><span class="p">[:,</span> <span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
                <span class="n">pool_emb</span> <span class="o">+=</span> <span class="n">cell_encode_x</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">pool</span> <span class="o">==</span> <span class="s1">&#39;mean&#39;</span><span class="p">:</span>
                    <span class="n">pool_emb</span> <span class="o">=</span> <span class="n">pool_emb</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="p">:</span>
            <span class="n">total_gpu</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">memory_allocated</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="o">.</span><span class="n">log</span><span class="p">({</span><span class="s1">&#39;GPU&#39;</span><span class="p">:</span> <span class="n">total_gpu</span><span class="p">})</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;get gene expression Done!&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pool_emb</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span></div>


<div class="viewcode-block" id="LoadScbert.get_cell_embedding">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.get_cell_embedding">[docs]</a>
    <span class="k">def</span> <span class="nf">get_cell_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">cell_emb_type</span><span class="o">=</span><span class="s1">&#39;cls&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Extracts cell embeddings using specified aggregation methods.</span>

<span class="sd">        Args:</span>
<span class="sd">            adata (AnnData): Single-cell data.</span>
<span class="sd">            cell_emb_type (str): Embedding aggregation type (&#39;cls&#39;, &#39;sum&#39;, or &#39;mean&#39;).</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: Cell embeddings as a NumPy array.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;start to get cell embedding!&quot;</span><span class="p">)</span>
        <span class="n">dataset</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_dataset</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">split_data</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">data_loader</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_dataloader</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">random_sample</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">(),</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">amp</span><span class="o">.</span><span class="n">autocast</span><span class="p">(</span><span class="n">enabled</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
            <span class="n">cell_embeddings</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">data_loader</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s1">&#39;get scbert cell embedding: &#39;</span><span class="p">):</span>
                <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
                <span class="n">cell_encode_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">return_encodings</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># [batch size, max_seq_len, dim]</span>
                <span class="k">if</span> <span class="n">cell_emb_type</span> <span class="o">==</span> <span class="s1">&#39;cls&#39;</span><span class="p">:</span>
                    <span class="n">cell_emb</span> <span class="o">=</span> <span class="n">cell_encode_x</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
                <span class="k">elif</span> <span class="n">cell_emb_type</span> <span class="o">==</span> <span class="s1">&#39;sum&#39;</span><span class="p">:</span>
                    <span class="n">cell_emb</span> <span class="o">=</span> <span class="n">cell_encode_x</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">cell_emb</span> <span class="o">=</span> <span class="n">cell_encode_x</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">cell_embeddings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell_emb</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="p">:</span>
            <span class="n">total_gpu</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">memory_allocated</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="o">.</span><span class="n">log</span><span class="p">({</span><span class="s1">&#39;GPU&#39;</span><span class="p">:</span> <span class="n">total_gpu</span><span class="p">})</span>
        <span class="n">cell_embeddings</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">cell_embeddings</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;end to get cell embedding!&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cell_embeddings</span></div>


<div class="viewcode-block" id="LoadScbert.get_embedding">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.get_embedding">[docs]</a>
    <span class="k">def</span> <span class="nf">get_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">emb_type</span><span class="p">,</span> <span class="n">adata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gene_ids</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Obtains embeddings for either genes, cells, or gene-expression profiles.</span>

<span class="sd">        Args:</span>
<span class="sd">            emb_type (str): Type of embedding (&#39;gene&#39;, &#39;cell&#39;, or &#39;gene-expression&#39;).</span>
<span class="sd">            adata (AnnData, optional): Single-cell data for cell or gene-expression embeddings.</span>
<span class="sd">            gene_ids (torch.Tensor, optional): Gene IDs for gene embeddings.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: Embedding data based on the requested type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">()</span>
        <span class="k">assert</span> <span class="n">emb_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;cell&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;gene-expression&#39;</span><span class="p">],</span> <span class="s1">&#39;Invalid embedding type, must be gene, cell or gene-expression&#39;</span>
        <span class="k">if</span> <span class="n">emb_type</span> <span class="o">==</span> <span class="s1">&#39;gene&#39;</span> <span class="ow">and</span> <span class="n">gene_ids</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;gene_ids must not be None if emb_type is gene!&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">emb_type</span> <span class="o">!=</span> <span class="s1">&#39;gene&#39;</span> <span class="ow">and</span> <span class="n">adata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;adata must not be None if emb_type is cell or gene-expression!&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">emb_type</span> <span class="o">==</span> <span class="s1">&#39;gene&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_embedding</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">emb_type</span> <span class="o">==</span> <span class="s1">&#39;cell&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cell_embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_emb_type</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">cell_emb_type</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_expression_embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span></div>


<div class="viewcode-block" id="LoadScbert.encoder">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scbert.LoadScbert.encoder">[docs]</a>
    <span class="k">def</span> <span class="nf">encoder</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">batch_data</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Encodes a batch of data using the SCBERT model.</span>

<span class="sd">        Args:</span>
<span class="sd">            batch_data (torch.Tensor): Batch of input data to encode.</span>

<span class="sd">        Returns:</span>
<span class="sd">            torch.Tensor: Encoded representations of the input data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cell_encode_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">(</span><span class="n">batch_data</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cell_encode_x</span></div>
</div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">biollm.utils.utils</span> <span class="kn">import</span> <span class="n">load_config</span>
    <span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

    <span class="n">config_file</span> <span class="o">=</span> <span class="s1">&#39;../config/embeddings/scbert_emb.toml&#39;</span>
    <span class="n">configs</span> <span class="o">=</span> <span class="n">load_config</span><span class="p">(</span><span class="n">config_file</span><span class="p">)</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">configs</span><span class="o">.</span><span class="n">input_file</span><span class="p">)</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">100</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">obj</span> <span class="o">=</span> <span class="n">LoadScbert</span><span class="p">(</span><span class="n">configs</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">obj</span><span class="o">.</span><span class="n">args</span><span class="p">)</span>
    <span class="k">if</span> <span class="s1">&#39;gene_name&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">:</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">obj</span><span class="o">.</span><span class="n">get_gene2idx</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">)</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">configs</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
    <span class="n">emb</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">get_embedding</span><span class="p">(</span><span class="n">obj</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">emb_type</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;embedding shape:&#39;</span><span class="p">,</span> <span class="n">emb</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">obj</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;/scbert_</span><span class="si">{</span><span class="n">obj</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">emb_type</span><span class="si">}</span><span class="s1">_emb.pk&#39;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">w</span><span class="p">:</span>
        <span class="n">pkl</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">emb</span><span class="p">,</span> <span class="n">w</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, BGI, Ping Qiu.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>