

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>biollm.base.load_scfoundation &mdash; BioLLM 0.1.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=a58bc63e"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            BioLLM
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../Cell_embedding.html">Evaluation of the clustering of the embed cell.</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../GRN.html">GRN</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Annotation.html">Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../drug_task.html">Drug Annotation Task Documentation</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">BioLLM Modules</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">References:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../download.html">Download the scFM pretrain model and configs</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">BioLLM</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">biollm.base.load_scfoundation</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for biollm.base.load_scfoundation</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># -*- coding:utf-8 -*-</span>
<span class="c1"># @FileName  :load_scFoundation</span>
<span class="c1"># @Time      :2024/3/20 15:28</span>
<span class="c1"># @Author    :Luni Hu</span>

<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">biollm.base.load_llm</span> <span class="kn">import</span> <span class="n">LoadLlm</span>
<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">issparse</span>
<span class="kn">from</span> <span class="nn">biollm.repo.scgpt.tokenizer.gene_tokenizer</span> <span class="kn">import</span> <span class="n">GeneVocab</span>
<span class="kn">from</span> <span class="nn">biollm.repo.scfoundation.load</span> <span class="kn">import</span> <span class="n">load_model_frommmf</span>
<span class="kn">from</span> <span class="nn">biollm.repo.scfoundation.get_embedding</span> <span class="kn">import</span> <span class="n">main_gene_selection</span><span class="p">,</span> <span class="n">embedding</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">biollm.repo.scfoundation.load</span> <span class="kn">import</span> <span class="n">gatherData</span><span class="p">,</span> <span class="n">getEncoerDecoderData</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">from</span> <span class="nn">torch.utils.data.distributed</span> <span class="kn">import</span> <span class="n">DistributedSampler</span>
<span class="kn">from</span> <span class="nn">torch.utils.data.sampler</span> <span class="kn">import</span> <span class="n">RandomSampler</span><span class="p">,</span> <span class="n">SequentialSampler</span>
<span class="kn">from</span> <span class="nn">torch.utils.data</span> <span class="kn">import</span> <span class="n">DataLoader</span>


<div class="viewcode-block" id="LoadScfoundation">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation">[docs]</a>
<span class="k">class</span> <span class="nc">LoadScfoundation</span><span class="p">(</span><span class="n">LoadLlm</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    LoadScfoundation class is a specialized loader for scFoundation model in the BioLLM framework.</span>
<span class="sd">    It initializes and manages a single-cell model, processes gene expression data, and generates gene and cell embeddings.</span>

<span class="sd">    Attributes:</span>
<span class="sd">    vocab : GeneVocab</span>
<span class="sd">        Gene vocabulary loaded from a specified file.</span>
<span class="sd">    model : torch.nn.Module</span>
<span class="sd">        The pretrained foundational model loaded from the specified file.</span>
<span class="sd">    config : dict</span>
<span class="sd">        Configuration dictionary for the model, including settings for the encoder and decoder.</span>
<span class="sd">    device : torch.device</span>
<span class="sd">        The computational device (CPU or GPU) where the model runs.</span>
<span class="sd">    gene2idx : dict</span>
<span class="sd">        Dictionary mapping gene symbols to indices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">args</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cfs_file</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">LoadScfoundation</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__int__</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">cfs_file</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">vocab</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_vocab</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_model</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gene2idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gene2idx</span><span class="p">()</span>

<div class="viewcode-block" id="LoadScfoundation.load_model">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.load_model">[docs]</a>
    <span class="k">def</span> <span class="nf">load_model</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads the foundational model and configuration from a specified file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: A tuple containing the pretrained foundational model (torch.nn.Module)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">model</span><span class="p">,</span> <span class="n">config</span> <span class="o">=</span> <span class="n">load_model_frommmf</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">model_file</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">key</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">model</span><span class="p">,</span> <span class="n">config</span></div>


<div class="viewcode-block" id="LoadScfoundation.get_embedding">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.get_embedding">[docs]</a>
    <span class="k">def</span> <span class="nf">get_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">emb_type</span><span class="p">,</span> <span class="n">adata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">gene_ids</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves the embedding for genes, cells, or gene expressions, depending on the specified type.</span>

<span class="sd">        Args:</span>
<span class="sd">            emb_type (str): Type of embedding to generate (&#39;gene&#39;, &#39;cell&#39;, &#39;gene-expression&#39;).</span>
<span class="sd">            adata (AnnData, optional): Annotated data object required for &#39;cell&#39; and &#39;gene-expression&#39; embeddings.</span>
<span class="sd">            gene_ids (list of int, optional): Gene IDs required for &#39;gene&#39; embedding.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: The computed embeddings as a NumPy array.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">biollm.utils.preprocess</span> <span class="kn">import</span> <span class="n">preprocess_adata</span>
        <span class="k">if</span> <span class="n">adata</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">preprocess_adata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">n_hvg</span> <span class="k">if</span> <span class="s1">&#39;n_hvg&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span> <span class="k">else</span> <span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">()</span>
        <span class="k">assert</span> <span class="n">emb_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;cell&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;gene-expression&#39;</span><span class="p">],</span> <span class="s1">&#39;Invalid embedding type, must be gene, cell or gene-expression&#39;</span>
        <span class="k">if</span> <span class="n">emb_type</span> <span class="o">==</span> <span class="s1">&#39;gene&#39;</span> <span class="ow">and</span> <span class="n">gene_ids</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;gene_ids must not be None if emb_type is gene!&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">emb_type</span> <span class="o">!=</span> <span class="s1">&#39;gene&#39;</span> <span class="ow">and</span> <span class="n">adata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;adata must not be None if emb_type is cell or gene-expression!&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">emb_type</span> <span class="o">==</span> <span class="s1">&#39;gene&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_embedding</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">emb_type</span> <span class="o">==</span> <span class="s1">&#39;cell&#39;</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_cell_embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_gene_expression_embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">emb</span></div>


<div class="viewcode-block" id="LoadScfoundation.get_gene_embedding">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.get_gene_embedding">[docs]</a>
    <span class="k">def</span> <span class="nf">get_gene_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Computes embeddings for specified genes using the model&#39;s positional embedding layer.</span>

<span class="sd">        Args:</span>
<span class="sd">            gene_ids (list of int): IDs of genes to generate embeddings for.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: Gene embeddings as a NumPy array.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;start to get gene embedding!&#39;</span><span class="p">)</span>
        <span class="n">emb</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">pos_emb</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="p">:</span>
            <span class="n">total_gpu</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">memory_allocated</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="o">.</span><span class="n">log</span><span class="p">({</span><span class="s1">&#39;GPU&#39;</span><span class="p">:</span> <span class="n">total_gpu</span><span class="p">})</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;start to get gene embedding!&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">emb</span><span class="p">(</span><span class="n">gene_ids</span><span class="p">)</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span></div>


<div class="viewcode-block" id="LoadScfoundation.get_gene2idx">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.get_gene2idx">[docs]</a>
    <span class="k">def</span> <span class="nf">get_gene2idx</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Retrieves the gene-to-index mapping from the vocabulary.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: Mapping of gene symbols to indices.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">vocab</span><span class="o">.</span><span class="n">get_stoi</span><span class="p">()</span></div>


<div class="viewcode-block" id="LoadScfoundation.load_vocab">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.load_vocab">[docs]</a>
    <span class="k">def</span> <span class="nf">load_vocab</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads gene vocabulary from a specified file.</span>

<span class="sd">        Returns:</span>
<span class="sd">            GeneVocab: Gene vocabulary object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">vocab</span> <span class="o">=</span> <span class="n">GeneVocab</span><span class="o">.</span><span class="n">from_file</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">vocab_file</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">vocab</span></div>


<div class="viewcode-block" id="LoadScfoundation.load_data">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.load_data">[docs]</a>
    <span class="k">def</span> <span class="nf">load_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_none_zore</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Loads gene expression data and applies sparse selection based on non-zero thresholding.</span>

<span class="sd">        Args:</span>
<span class="sd">            adata (AnnData, optional): Annotated data object containing gene expression data.</span>
<span class="sd">            max_none_zore (int, optional): Threshold for non-zero values in gene expression data.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: Processed gene expression data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">adata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">input_file</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">col</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">issparse</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">):</span>
            <span class="n">gexpr_feature</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gexpr_feature</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">max_none_zore</span><span class="p">:</span>
            <span class="n">none_zero</span> <span class="o">=</span> <span class="n">gexpr_feature</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="n">none_zero_num</span> <span class="o">=</span> <span class="n">none_zero</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">none_zero_num</span> <span class="o">&gt;</span> <span class="n">max_none_zore</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">index</span><span class="p">:</span>
                <span class="n">none_zero_index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">none_zero</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">none_zero_index</span><span class="p">)</span>
                <span class="n">mask_num</span> <span class="o">=</span> <span class="n">none_zero_num</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">-</span> <span class="n">max_none_zore</span>
                <span class="n">mask_index</span> <span class="o">=</span> <span class="n">none_zero_index</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span> <span class="n">mask_num</span><span class="p">]</span>
                <span class="n">gexpr_feature</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">mask_index</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">gexpr_feature</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gexpr_feature</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">idx</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">col</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;covert gene feature into 19264&#39;</span><span class="p">)</span>
        <span class="n">gene_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">vocab</span><span class="o">.</span><span class="n">get_stoi</span><span class="p">()</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">gexpr_feature</span> <span class="o">=</span> <span class="n">gexpr_feature</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">gexpr_feature</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)]</span>
        <span class="n">gexpr_feature</span><span class="p">,</span> <span class="n">to_fill_columns</span><span class="p">,</span> <span class="n">var</span> <span class="o">=</span> <span class="n">main_gene_selection</span><span class="p">(</span><span class="n">gexpr_feature</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">gexpr_feature</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">19264</span>
        <span class="k">return</span> <span class="n">gexpr_feature</span></div>


<div class="viewcode-block" id="LoadScfoundation.get_dataloader">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.get_dataloader">[docs]</a>
    <span class="k">def</span> <span class="nf">get_dataloader</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dataset</span><span class="p">,</span> <span class="n">batch_size</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ddp_train</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">drop_last</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">0</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructs a DataLoader for efficient batched data processing.</span>

<span class="sd">        Args:</span>
<span class="sd">            dataset (torch.utils.data.Dataset): Dataset to be loaded.</span>
<span class="sd">            batch_size (int): Number of samples per batch.</span>
<span class="sd">            shuffle (bool, optional): Whether to shuffle the data.</span>
<span class="sd">            ddp_train (bool, optional): Whether to use distributed data parallel (DDP) training.</span>
<span class="sd">            drop_last (bool, optional): Whether to drop the last incomplete batch.</span>
<span class="sd">            num_workers (int, optional): Number of subprocesses to use for data loading.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataLoader: DataLoader instance for the dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">ddp_train</span><span class="p">:</span>
            <span class="n">sampler</span> <span class="o">=</span> <span class="n">DistributedSampler</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sampler</span> <span class="o">=</span> <span class="n">SequentialSampler</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">shuffle</span> <span class="k">else</span> <span class="n">RandomSampler</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
        <span class="n">data_loader</span> <span class="o">=</span> <span class="n">DataLoader</span><span class="p">(</span>
            <span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">,</span>
            <span class="n">batch_size</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
            <span class="n">drop_last</span><span class="o">=</span><span class="n">drop_last</span><span class="p">,</span>
            <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">,</span>
            <span class="n">sampler</span><span class="o">=</span><span class="n">sampler</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">data_loader</span></div>


<div class="viewcode-block" id="LoadScfoundation.make_encoder_input">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.make_encoder_input">[docs]</a>
    <span class="k">def</span> <span class="nf">make_encoder_input</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gexpr_feature</span><span class="p">,</span> <span class="n">tgthighres</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructs encoder input features by combining gene expression data and target resolution.</span>

<span class="sd">        Args:</span>
<span class="sd">            gexpr_feature (pd.DataFrame): Gene expression feature data.</span>
<span class="sd">            tgthighres (str): Target resolution information.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: Prepared input features for the encoder.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">gexpr_feature</span><span class="o">.</span><span class="n">values</span>
        <span class="n">totalcount</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">tgthighres</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;f&#39;</span><span class="p">:</span>
            <span class="n">pretrain_gene_x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">totalcount</span> <span class="o">*</span> <span class="nb">float</span><span class="p">(</span><span class="n">tgthighres</span><span class="p">[</span><span class="mi">1</span><span class="p">:])),</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">totalcount</span><span class="p">)],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">tgthighres</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;a&#39;</span><span class="p">:</span>
            <span class="n">pretrain_gene_x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">totalcount</span> <span class="o">+</span> <span class="nb">float</span><span class="p">(</span><span class="n">tgthighres</span><span class="p">[</span><span class="mi">1</span><span class="p">:])),</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">totalcount</span><span class="p">)],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">tgthighres</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;t&#39;</span><span class="p">:</span>
            <span class="n">pretrain_gene_x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">full_like</span><span class="p">(</span><span class="n">totalcount</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">(</span><span class="n">tgthighres</span><span class="p">[</span><span class="mi">1</span><span class="p">:])),</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">totalcount</span><span class="p">)],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;tgthighres must be start with f, a or t&#39;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pretrain_gene_x</span></div>


<div class="viewcode-block" id="LoadScfoundation.get_gene_expression_embedding">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.get_gene_expression_embedding">[docs]</a>
    <span class="k">def</span> <span class="nf">get_gene_expression_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">pool</span><span class="o">=</span><span class="s1">&#39;mean&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Obtains gene expression embeddings by pooling the model’s cell-level embeddings.</span>

<span class="sd">        Args:</span>
<span class="sd">            adata (AnnData): Single-cell data.</span>
<span class="sd">            pool (str): Pooling method, either &#39;mean&#39; or &#39;max&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: Gene expression embeddings as a NumPy array.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_data</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_none_zore</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">max_none_zero_num</span><span class="p">)</span>
        <span class="n">pretrain_gene_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_encoder_input</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">tgthighres</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">to_final</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">(),</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">amp</span><span class="o">.</span><span class="n">autocast</span><span class="p">():</span>
            <span class="n">pool_emb</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">gene2idx</span><span class="p">),</span> <span class="mi">512</span><span class="p">))</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pretrain_gene_x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">batch_size</span><span class="p">)):</span>
                <span class="n">x</span> <span class="o">=</span> <span class="n">pretrain_gene_x</span><span class="p">[</span><span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="p">:]</span>
                <span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
                <span class="n">encoder_data</span><span class="p">,</span> <span class="n">encoder_position_gene_ids</span><span class="p">,</span> <span class="n">encoder_data_padding</span><span class="p">,</span> <span class="n">encoder_labels</span><span class="p">,</span> <span class="n">decoder_data</span><span class="p">,</span> <span class="n">decoder_data_padding</span><span class="p">,</span> <span class="n">new_data_raw</span><span class="p">,</span> <span class="n">data_mask_labels</span><span class="p">,</span> <span class="n">decoder_position_gene_ids</span> <span class="o">=</span> <span class="n">getEncoerDecoderData</span><span class="p">(</span>
                <span class="n">x</span><span class="o">.</span><span class="n">float</span><span class="p">(),</span> <span class="n">x</span><span class="o">.</span><span class="n">float</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">)</span>
                <span class="n">out</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">forward</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">encoder_data</span><span class="p">,</span> <span class="n">padding_label</span><span class="o">=</span><span class="n">encoder_data_padding</span><span class="p">,</span>
                                    <span class="n">encoder_position_gene_ids</span><span class="o">=</span><span class="n">encoder_position_gene_ids</span><span class="p">,</span>
                                    <span class="n">encoder_labels</span><span class="o">=</span><span class="n">encoder_labels</span><span class="p">,</span>
                                    <span class="n">decoder_data</span><span class="o">=</span><span class="n">decoder_data</span><span class="p">,</span>
                                    <span class="n">mask_gene_name</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                    <span class="n">mask_labels</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                    <span class="n">decoder_position_gene_ids</span><span class="o">=</span><span class="n">decoder_position_gene_ids</span><span class="p">,</span>
                                    <span class="n">decoder_data_padding_labels</span><span class="o">=</span><span class="n">decoder_data_padding</span><span class="p">,</span>
                                    <span class="p">)</span>
                <span class="n">out</span> <span class="o">=</span> <span class="n">out</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">19264</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">contiguous</span><span class="p">()</span>
                <span class="n">pool_emb</span> <span class="o">+=</span> <span class="n">out</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">pool</span> <span class="o">==</span> <span class="s1">&#39;mean&#39;</span><span class="p">:</span>
                    <span class="n">pool_emb</span> <span class="o">=</span> <span class="n">pool_emb</span> <span class="o">/</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="p">:</span>
            <span class="n">total_gpu</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">memory_allocated</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="o">.</span><span class="n">log</span><span class="p">({</span><span class="s1">&#39;GPU&#39;</span><span class="p">:</span> <span class="n">total_gpu</span><span class="p">})</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;get gene expression Done!&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pool_emb</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span></div>


<div class="viewcode-block" id="LoadScfoundation.get_cell_embedding">
<a class="viewcode-back" href="../../../api.html#biollm.base.load_scfoundation.LoadScfoundation.get_cell_embedding">[docs]</a>
    <span class="k">def</span> <span class="nf">get_cell_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">pool</span><span class="o">=</span><span class="s1">&#39;max&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Obtains cell embeddings by processing the gene expression data through the model.</span>

<span class="sd">        Args:</span>
<span class="sd">            adata (AnnData): Single-cell data.</span>
<span class="sd">            pool (str): Pooling method, either &#39;all&#39; or &#39;max&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.ndarray: Cell embeddings as a NumPy array.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_data</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_none_zore</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">max_none_zero_num</span><span class="p">)</span>
        <span class="n">pretrain_gene_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_encoder_input</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">tgthighres</span><span class="p">)</span>
        <span class="n">cell_embeddings</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">(),</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">amp</span><span class="o">.</span><span class="n">autocast</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">pretrain_gene_x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">batch_size</span><span class="p">)):</span>
                <span class="n">x</span> <span class="o">=</span> <span class="n">pretrain_gene_x</span><span class="p">[</span><span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="o">+</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="p">:]</span>
                <span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
                <span class="n">value_labels</span> <span class="o">=</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span>
                <span class="n">data_gene_ids</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">19266</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">device</span><span class="p">)</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
                <span class="n">x</span><span class="p">,</span> <span class="n">x_padding</span> <span class="o">=</span> <span class="n">gatherData</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">value_labels</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;pad_token_id&#39;</span><span class="p">])</span>
                <span class="n">position_gene_ids</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">gatherData</span><span class="p">(</span><span class="n">data_gene_ids</span><span class="p">,</span> <span class="n">value_labels</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">config</span><span class="p">[</span><span class="s1">&#39;pad_token_id&#39;</span><span class="p">])</span>
                <span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">token_emb</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">float</span><span class="p">(),</span> <span class="n">output_weight</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="n">position_emb</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">pos_emb</span><span class="p">(</span><span class="n">position_gene_ids</span><span class="p">)</span>
                <span class="n">x</span> <span class="o">+=</span> <span class="n">position_emb</span>
                <span class="n">geneemb</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">encoder</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">x_padding</span><span class="p">)</span>
                <span class="n">geneemb1</span> <span class="o">=</span> <span class="n">geneemb</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
                <span class="n">geneemb2</span> <span class="o">=</span> <span class="n">geneemb</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="p">:]</span>
                <span class="n">geneemb3</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">geneemb</span><span class="p">[:,</span> <span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="p">:],</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">geneemb4</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">geneemb</span><span class="p">[:,</span> <span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="p">:],</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">pool</span> <span class="o">==</span> <span class="s1">&#39;all&#39;</span><span class="p">:</span>
                    <span class="n">geneembmerge</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">geneemb1</span><span class="p">,</span> <span class="n">geneemb2</span><span class="p">,</span> <span class="n">geneemb3</span><span class="p">,</span> <span class="n">geneemb4</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">pool</span> <span class="o">==</span> <span class="s1">&#39;max&#39;</span><span class="p">:</span>
                    <span class="n">geneembmerge</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">geneemb</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;pool_type must be all or max&#39;</span><span class="p">)</span>
                <span class="n">cell_embeddings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">geneembmerge</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="p">:</span>
            <span class="n">total_gpu</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">memory_allocated</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="mi">1024</span><span class="o">**</span><span class="mi">3</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">wandb</span><span class="o">.</span><span class="n">log</span><span class="p">({</span><span class="s1">&#39;GPU&#39;</span><span class="p">:</span> <span class="n">total_gpu</span><span class="p">})</span>
        <span class="n">cell_embeddings</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">cell_embeddings</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;end to get cell embedding!&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cell_embeddings</span></div>
</div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">biollm.utils.utils</span> <span class="kn">import</span> <span class="n">load_config</span>
    <span class="kn">import</span> <span class="nn">pickle</span> <span class="k">as</span> <span class="nn">pkl</span>
    <span class="kn">import</span> <span class="nn">os</span>
    <span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

    <span class="n">config_file</span> <span class="o">=</span> <span class="s1">&#39;../../tutorials/zero-shot/configs/scfoundation_cell_emb.toml&#39;</span>
    <span class="n">configs</span> <span class="o">=</span> <span class="n">load_config</span><span class="p">(</span><span class="n">config_file</span><span class="p">)</span>

    <span class="n">obj</span> <span class="o">=</span> <span class="n">LoadScfoundation</span><span class="p">(</span><span class="n">configs</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">obj</span><span class="o">.</span><span class="n">args</span><span class="p">)</span>

    <span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">configs</span><span class="o">.</span><span class="n">input_file</span><span class="p">)</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:</span><span class="mi">1000</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">obj</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">configs</span><span class="o">.</span><span class="n">device</span><span class="p">)</span>
    <span class="n">emb</span> <span class="o">=</span> <span class="n">obj</span><span class="o">.</span><span class="n">get_embedding</span><span class="p">(</span><span class="n">configs</span><span class="o">.</span><span class="n">emb_type</span><span class="p">,</span> <span class="n">gene_ids</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;embedding shape:&#39;</span><span class="p">,</span> <span class="n">emb</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">configs</span><span class="o">.</span><span class="n">output_dir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">configs</span><span class="o">.</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">obj</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">output_dir</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;/scfoundation_</span><span class="si">{</span><span class="n">obj</span><span class="o">.</span><span class="n">args</span><span class="o">.</span><span class="n">emb_type</span><span class="si">}</span><span class="s1">_emb.pk&#39;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">w</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;gene_names&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">obj</span><span class="o">.</span><span class="n">get_gene2idx</span><span class="p">()</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="s1">&#39;gene_emb&#39;</span><span class="p">:</span> <span class="n">emb</span><span class="p">}</span>
        <span class="n">pkl</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">emb</span><span class="p">,</span> <span class="n">w</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, BGI, Ping Qiu.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>